### PR TITLE
feat(agent): persist to statefile at a regular interval

### LINKED ISSUE #14566: update state more frequently incase telegraf is suddenly shutdown
### Use Case

Currently state is only stored once at the end of a telegraf run. In cases where telegraf fails catastrophically, this means lost of state. 

### Expected behavior

When the telegraf container has read from a large log file via tail plugin and is suddenly deleted without a safe shutdown, state is recorded.

### Actual behavior

When the telegraf container has read from a large log file via tail plugin and is suddenly deleted without a safe shutdown, state is not recorded.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by dayvar14**
## Summary

State is now stored at each gather for each plugin individually.
Screenshot 2024-01-11 at 11 14 09 AM

State is restored correctly if telegraf fails catastrophically.
Screenshot 2024-01-11 at 11 19 46 AM

Change shouldn't affect any current implementation of state usage as its simply writing to the state more frequently.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #14566

### CODE REVIEW

**Comment by powersj** [comment]
Hi @dayvar14,

Thanks for putting up the PR. We have a few concerns about this change. This is a scenario that we talked about in depth when we first added the statefile feature. Catching the state during a crash was ultimately not included in the first pass.

There are two main points that need to be resolved:

First, this needs to be opt-in and not the default behavior.

Second, doing this for every gather cycle could also be I/O heavy depending on your deployment and your interval setting. Possibly having a config option for how often this is done may be necessary.

Keep in mind that not all plugins even have a gather cycle as well. While you are doing this in the agent for each interval, knowing how this operates on other plugins types, like service input plugins that listen for metrics would be good to ensure is tested.

Thanks

File: cmd/telegraf/agent.conf [hunk 1]
```diff
@@ -94,3 +94,14 @@
   ## stateful plugins on termination of Telegraf. If the file exists on start,
   ## the state in the file will be restored for the plugins.
   # statefile = ""
+
+  ## If set to true, the agent will persist it's state at a regular interval set
+  ## by statefile_persist_interval.
+  # enable_statefile_persist_interval = false
+
+  ## The interval of time at which the state is peristed into the statefile.
+  # statefile_persist_interval = "10m"
```
Comment (Hipska) [inline cmd/telegraf/agent.conf:103]: What if we combine both into one config option?

```suggestion
  ## The interval of time at which the state is peristed into the statefile.
  ## Default 0s means the agent will not persist the state at intervals.
  ## example: statefile_persist_interval = "10m"
  # statefile_persist_interval = "0s"
```

Comment (dayvar14) [reply cmd/telegraf/agent.conf:103]: Makes sense. The amount of options was getting a bit lengthy anyways.

Comment (dayvar14) [reply cmd/telegraf/agent.conf:103]: Although it could lead to a bit of confusion as the behavior for the other 'intervals' dont disable if set to zero.

Comment (Hipska) [reply cmd/telegraf/agent.conf:103]: There is also no option to disable those, they are a core part of the working of telegraf.

Comment (dayvar14) [reply cmd/telegraf/agent.conf]: Should be resolved as of this [commit](https://github.com/influxdata/telegraf/pull/14561/commits/f835b958fb0c373c5e7e6751e45d50227ee59c7c)

File: cmd/telegraf/agent.conf [hunk 2]
```diff
@@ -94,3 +94,11 @@
   ## stateful plugins on termination of Telegraf. If the file exists on start,
   ## the state in the file will be restored for the plugins.
   # statefile = ""
+
+  ## The interval of time at which the state is peristed into the statefile.
+  ## When set to 0, persist interval is disabled.
+  # statefile_persist_interval = "0h"
```
Comment (Hipska) [inline cmd/telegraf/agent.conf:100]:
```suggestion
  ## When set to 0s, persist interval is disabled.
  ## Example: statefile_persist_interval = "10m"
  # statefile_persist_interval = "0s"
```

File: cmd/telegraf/telegraf.go [hunk 3]
```diff
@@ -278,6 +278,10 @@ func (t *Telegraf) runAgent(ctx context.Context, c *config.Config, reloadConfig
 		return fmt.Errorf("agent flush_interval must be positive; found %v", c.Agent.Interval)
 	}
 
+	if int64(c.Agent.PersistInterval) < 0 {
+		return fmt.Errorf("agent statefile_persist_interval must be positive or zero; found %v", c.Agent.Interval)
```
Comment (Hipska) [inline cmd/telegraf/telegraf.go:282]:
```suggestion
		return fmt.Errorf("agent statefile_persist_interval must be positive or zero; found %v", c.Agent.PersistInterval)
```

File: cmd/telegraf/agent.conf [hunk 4]
```diff
@@ -94,3 +94,11 @@
...
+
+  ## Rounds persist interval to 'interval'
+  ## ie, if interval="10s" then always persist state on :00, :10, :20, etc.
```
Comment (Hipska) [inline cmd/telegraf/agent.conf:103]:
```suggestion
  ## Rounds persist interval to 'statefile_persist_interval'
  ## ie, if statefile_persist_interval="10m" then always persist state on :00, :10, :20, etc.
```

**Comment by dayvar14** [comment]
Ran the tests in the same test described in the issue and 
* the state is persisting correctly.
* Persist loop stopping correctly
* persist loop not starting if no interval not enabled in conf
Screenshot 2024-01-18 at 4 41 11 AM

File: config/config.go [hunk 5]
```diff
@@ -258,6 +260,14 @@ type AgentConfig struct {
 	// the state in the file will be restored for the plugins.
 	Statefile string `toml:"statefile"`
 
+	// PersistInterval The interval in which the state will be stored into the
+	// statefile. When set to 0, persist interval is disabled.
+	PersistInterval Duration `toml:"statefile_persist_interval"`
+
+	// RoundPersistInterval rounds persist interval to 'interval'.
+	//     ie, if Interval=10s then always persist state on :00, :10, :20, etc.
```
Comment (Hipska) [inline config/config.go:268]:
```suggestion
	// RoundPersistInterval rounds persist interval to 'PersistInterval'.
	//     ie, if PersistInterval=10m then always persist state on :00, :10, :20, etc.
```

File: docs/CONFIGURATION.md [hunk 6]
```diff
@@ -332,6 +332,14 @@ The agent table configures Telegraf and the defaults used across all plugins.
   stateful plugins on termination of Telegraf. If the file exists on start,
   the state in the file will be restored for the plugins.
 
+- **statefile_persist_interval**:
+  The interval of time at which the state is peristed into the statefile. When
+  set to 0, persist interval is disabled.
+
+- **round_statefile_persist_interval**:
+  Rounds persist interval to 'interval'
+  ie, if interval="10s" then always persist state on :00, :10, :20, etc.
```
Comment (Hipska) [inline docs/CONFIGURATION.md:341]:
```suggestion
  Rounds persist interval to 'statefile_persist_interval'
  ie, if statefile_persist_interval ="10m" then always persist state on :00, :10, :20, etc.
```

**Review by Hipska — CHANGES_REQUESTED** [review]

File: docs/CONFIGURATION.md (hunk 6 above)
Comment (dayvar14) [reply docs/CONFIGURATION.md]: Should be resolved as of this [commit](https://github.com/influxdata/telegraf/pull/14561/commits/35de14b84db306904d5376dbea7f3f27ece47a95)

File: config/config.go (hunk 5 above)
Comment (dayvar14) [reply config/config.go]: Should be resolved as of this [commit](https://github.com/influxdata/telegraf/pull/14561/commits/35de14b84db306904d5376dbea7f3f27ece47a95)

File: cmd/telegraf/agent.conf (hunk 4 above)
Comment (dayvar14) [reply cmd/telegraf/agent.conf]: Should be resolved as of this [commit](https://github.com/influxdata/telegraf/pull/14561/commits/35de14b84db306904d5376dbea7f3f27ece47a95)

File: cmd/telegraf/agent.conf (hunk 2 above)
Comment (dayvar14) [reply cmd/telegraf/agent.conf]: Should be resolved as of this [commit](https://github.com/influxdata/telegraf/pull/14561/commits/35de14b84db306904d5376dbea7f3f27ece47a95)

File: cmd/telegraf/agent.conf [hunk 7]
```diff
@@ -94,3 +94,12 @@
   ## stateful plugins on termination of Telegraf. If the file exists on start,
   ## the state in the file will be restored for the plugins.
   # statefile = ""
+
```
Comment (Hipska) [inline cmd/telegraf/agent.conf:97]: I did not suggest to remove that line, please revert.
```suggestion

  ## The interval of time at which the state is peristed into the statefile.
```

**Review by Hipska — CHANGES_REQUESTED** [review]

**Review by Hipska — APPROVED** [review]

**Comment by dayvar14** [comment]
Running this at scale, we are running into an issue where the state is getting overwritten to be an empty state

The state is written, but as noted by the cat at the end of the following, it ends up being empty.
```
root@pod-200100:/nas/log/cursors# cat telegraf_state
{}root@pod-200100:/nas/log/cursors#
root@pod-200100:/nas/log/cursors#
root@pod-200100:/nas/log/cursors# cat telegraf_state
{"561d8a29fded6f3c0f59d9f4a06e56eee3e07a1dbb30e4097c092dee275936e6":"e30=","61c155d4dbf6f6840e9c5ce48fee7b6e11a8b833ef1260215eba072654a541af":"eyIvbmFzL2xvZy9uZ2lueC9mbHVlbnRkX2FjY2Vzc19sb2ciOjcxOTA1MzQ0fQ==","7b42e8762e30de6deb7c9c023a5d69997f7d083d0c813b1a689b450210ab2538":"eyIvbmFzL2xvZy9uZ2lueC9hY2Nlc3MuanNvbiI6MH0=","cd5e9742d4b246f2ad3e79a0fd7ac9f44204c494cbb50a4d2ea9a3d8a76a5856":"eyIvbmFzL2xvZy9teXNxbC9teXNxbC1zbG93LmxvZyI6OTgxfQ=="}root@pod-200100:/nas/log/cursors# stat telegraf_state
  File: telegraf_state
  Size: 445             Blocks: 8          IO Block: 4096   regular file
Device: 840h/2112d      Inode: 131080      Links: 1
Access: (0666/-rw-rw-rw-)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-01-18 17:54:16.228502741 +0000
Modify: 2024-01-18 17:54:00.004525428 +0000
Change: 2024-01-18 17:54:00.004525428 +0000
 Birth: -
root@pod-200100:/nas/log/cursors# date
Thu Jan 18 17:54:23 UTC 2024
root@pod-200100:/nas/log/cursors# date
Thu Jan 18 17:54:28 UTC 2024
root@pod-200100:/nas/log/cursors# date
Thu Jan 18 17:54:29 UTC 2024
root@pod-200100:/nas/log/cursors# date
Thu Jan 18 17:54:31 UTC 2024
root@pod-200100:/nas/log/cursors# cat telegraf_state
{}root@pod-200100:/nas/log/cursors#
root@pod-200100:/nas/log/cursors# 
```

**Comment by srebhan** [comment]
@dayvar14 thanks for your contribution! Quickly reading over the code and one concern is that you now introduce a potentially racy operations. That is, when you `GetState()` from the plugins, this should not interfere with e.g. `Gather()`  or `Write()` operations, so I guess we need some locking here.

My other concern is that you do not get a consistent state across all Telegraf, e.g. if your input is collected just before some data arrived but your output is collects just after the data was sent the disk state is not a snapshot... This is a well known issue in backup and might be "OK" for people, but we need to be clear on this in the documentation! Alternatively, we need a "global lock" to get a persistent state but I'm not sure we want to go down that rabbit-hole...

**Comment by dayvar14** [comment]
@srebhan by interfere do you mean that the access of the plugin state should be protected by a mutex lock? Or do the various threads using the input "Gather" "GetState" and "Write" functions need to be protected from using those operations concurrently?

**Comment by dayvar14** [comment]
```
2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access_log.txt"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 981 for "/nas/log/mysql/mysql-slow.log"
2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/mysql/mysql-slow.log"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access.json"
2024-01-18T18:02:35Z D! [agent] Input channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z I! [agent] Hang on, flushing any cached metrics before shutdown
2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z D! [outputs.influxdb] Wrote batch of 274 metrics in 77.130676ms
2024-01-18T18:02:35Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z D! [outputs.bigquery] Wrote batch of 582 metrics in 103.440871ms
2024-01-18T18:02:35Z D! [outputs.bigquery] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z I! [agent] Stopping running outputs
2024-01-18T18:02:35Z D! [agent] Persisting plugin states
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access_log.txt"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/mysql/mysql-slow.log"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
2024-01-18T18:02:35Z D! [agent] Stopped Successfully2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access_log.txt"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 981 for "/nas/log/mysql/mysql-slow.log"
2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/mysql/mysql-slow.log"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access.json"
2024-01-18T18:02:35Z D! [agent] Input channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z D! [agent] Processor channel closed
2024-01-18T18:02:35Z I! [agent] Hang on, flushing any cached metrics before shutdown
2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z D! [outputs.influxdb] Wrote batch of 274 metrics in 77.130676ms
2024-01-18T18:02:35Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z D! [outputs.bigquery] Wrote batch of 582 metrics in 103.440871ms
2024-01-18T18:02:35Z D! [outputs.bigquery] Buffer fullness: 0 / 10000 metrics
2024-01-18T18:02:35Z I! [agent] Stopping running outputs
2024-01-18T18:02:35Z D! [agent] Persisting plugin states
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access_log.txt"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/mysql/mysql-slow.log"
2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
2024-01-18T18:02:35Z D! [agent] Stopped Successfully
```
Another example of the state being written to, but then the offsets being overwritten at the end of a reset.

**Comment by srebhan** [comment]
@dayvar14 both if we really want a _globally consistent_ state. However, the minimum is to make sure that `GetState()` is thread-safe and returns a **consistent state for that plugin instance**! E.g. if you have two values in the state those might be updated in the `Gather()` (or in the listener or whatever) of the plugin while we are calling `GetState()`...

**Comment by dayvar14** [comment]
> ```
> 2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access_log.txt"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 981 for "/nas/log/mysql/mysql-slow.log"
> 2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/mysql/mysql-slow.log"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
> 2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access.json"
> 2024-01-18T18:02:35Z D! [agent] Input channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z I! [agent] Hang on, flushing any cached metrics before shutdown
> 2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z D! [outputs.influxdb] Wrote batch of 274 metrics in 77.130676ms
> 2024-01-18T18:02:35Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z D! [outputs.bigquery] Wrote batch of 582 metrics in 103.440871ms
> 2024-01-18T18:02:35Z D! [outputs.bigquery] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z I! [agent] Stopping running outputs
> 2024-01-18T18:02:35Z D! [agent] Persisting plugin states
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access_log.txt"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/mysql/mysql-slow.log"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
> 2024-01-18T18:02:35Z D! [agent] Stopped Successfully2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access_log.txt"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 981 for "/nas/log/mysql/mysql-slow.log"
> 2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/mysql/mysql-slow.log"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
> 2024-01-18T18:02:35Z D! [inputs.tail] Tail removed for "/nas/log/nginx/access.json"
> 2024-01-18T18:02:35Z D! [agent] Input channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z D! [agent] Processor channel closed
> 2024-01-18T18:02:35Z I! [agent] Hang on, flushing any cached metrics before shutdown
> 2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z D! [outputs.cloud_pubsub] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z D! [outputs.influxdb] Wrote batch of 274 metrics in 77.130676ms
> 2024-01-18T18:02:35Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z D! [outputs.bigquery] Wrote batch of 582 metrics in 103.440871ms
> 2024-01-18T18:02:35Z D! [outputs.bigquery] Buffer fullness: 0 / 10000 metrics
> 2024-01-18T18:02:35Z I! [agent] Stopping running outputs
> 2024-01-18T18:02:35Z D! [agent] Persisting plugin states
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access_log.txt"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/mysql/mysql-slow.log"
> 2024-01-18T18:02:35Z D! [inputs.tail] Recording offset 0 for "/nas/log/nginx/access.json"
> 2024-01-18T18:02:35Z D! [agent] Stopped Successfully
> ```
> 
> Another example of the state being written to, but then the offsets being overwritten at the end of a reset.

After reviewing the code with my colleagues (@noahneedscoffee and @ericmiller-wpengine), we discovered that when the tails are terminated at the Stop() function of Tail, the offsets are read as zero when the persister calls the final GetState() method. We resolved this issue by simply setting the tails map to an empty struct and adding a conditional statement in GetState. I will provide an update once we identify the source of the persister corrupt written state bug and conduct stress tests with a high volume of logs to be read from.

**Comment by dayvar14** [comment]
This change was stress tested with 3 tail plugins ingesting 100000 logs a second. State writes every 30 seconds. Was also tested in our environment and had correctly restored from state when restarted.

**Review by Hipska — APPROVED** [review]

**Comment by dayvar14** [comment]
@srebhan @powersj Is there any other concerns with getting this pr merged?

**Comment by powersj** [comment]
@dayvar14,

My apologies, I owe you a response. Looks like you have tested this thoroughly, and I have no doubt resolved it for the tail plugin. I think there was concern around other plugins that currently implement the state and ensuring we have a consistent mechanism.

@srebhan is that your current thinking as well?

**Comment by dayvar14** [comment]
@powersj No worries. I understand now the concern a bit better now.

I only found four plugins that are stateful at the moment. inputs.tail, inputs.docker_log, inputs.win_eventlog, and processors.dedup.

Are you considering as a solution that...
1. just locks around the states for each plugin that are stateful.
2. a more aggressive refactor of StatefulPlugins that is better suited for the persistance loop perhaps even a separate interface for plugins that support the persistance loop and this change only affect the behavior of the tail plugin specifically.

Or do you have another approach in mind?

**Comment by srebhan** [comment]
@dayvar14 my minimal request would be that the plugin's state is **consistent** when stored on disk. For the current implementation that is the case as all Gather operations and all dataflows have been stopped, so Telegraf is in a "frozen" state when we collect the states. Here an example for my concerns: Imagine we do tail 4 files and use a file output to save the results for illustration. Assume further that every file provides 32 bytes corresponding to 1 metric per "write" The following is fully possible even with locks around the state:

```
- file 1 is read and offset is set to 32 -> state(32, 0, 0, 0)
- file 2 is read and offset is set to 32 -> state(32, 32, 0, 0)
- file 3 is read and offset is set to 32 -> state(32, 32, 32, 0)
- file 4 is read and offset is set to 32 -> state(32, 32, 32, 32)
-> write is called with 4 metrics
- file 3 is read and offset is set to 32 -> state(32, 32, 64, 32)
- file 1 is read and offset is set to 32 -> state(64, 32, 64, 32)
* state-persistence is triggered for the plugin and stored state(64, 32, 64, 32)
**crash**
```
Now we lost the last two metrics in file 3 and file 1, because the state says we did read them already, but they are not written at the output!
In a non-event-based plugin (tail triggers state-changes on file changes) the state might be modified in the gather loop e.g. when gathering multiple servers, but in the middle of that loop the state might be persisted, so your state represents some intermediate gather state and you will have no chance to know at which point of the gather process it was taken.

One possible solution would be double buffering. It will not fix the lost-metric example above, but it might guarantee we do have a clear gather-state e.g. by always returning the state before the actual gather...

@dayvar14 does that make any sense?

**Comment by dayvar14** [comment]
I see @srebhan. 

From what I understand, as long as theres correct locking around state for stateful non-service plugins, this should only affect service plugins.

So I can update inputs.docker_log, inputs.win_eventlog, and processors.dedup accordingly.

However, for the tail input plugin specifically (which is a service plugin), I think there might be a slight limitation with the tail implementation `influxdata/tail` in use currently.

InfluxData's [implementation of tail](https://github.com/influxdata/tail/blob/19b97bffd9780cf48367e5df276b38e31c04c4e1/tail.go#L28-L32) does not output the line index of the currently read line. 

Meaning we are reliant on the tail.Tell(), which is noted specifically in the docs that its not very accurate and its very possible a new line was received by the time the line was processed.
https://github.com/influxdata/tail/blob/19b97bffd9780cf48367e5df276b38e31c04c4e1/tail.go#L139-L143

I do see some implementations of [tail that do this](https://pkg.go.dev/github.com/nxadm/tail#Line), so it may be possible to either use this or update the current implementation to support this.

If we do that, or if we are fine with the limitations of tail.Tell(), buffering that state and writing only when the metrics are delivered seems like a good potential solution.

Again thanks for the feedback. Sorry if i'm taking a bit of your time 😅

**Comment by powersj** [comment]
hi dayvar14,

> InfluxData's [implementation of tail](https://github.com/influxdata/tail/blob/19b97bffd9780cf48367e5df276b38e31c04c4e1/tail.go#L28-L32) does not output the line index of the currently read line.

For the tail plugin, does that provide something different than the offset used today?

As Sven was mentioning, when we start gathering state at an interval it may not be consistent with messages that are successfully written. The offset may have advanced, but the metric not written. If a crash happens before writing, this leads to you missing data and having small gaps in your state when you recover. There is a similar concern with the Windows Event Log and its use of a bookmark.

As Sven and I have reviewed this we are recalling why we did not originally add periodic state tracking. There are enough moving parts to make this racy and inconsistent. I would like to avoid take a PR that we know will result in this type of behavior for now.

**Comment by dayvar14** [comment]
I updated the following the following plugins win_eventlog, docker_log, and dedup plugins. Hopefully this alleviates your concerns @powersj. I do understand the wants. 

Win_eventlog - only updates the state at the end of a gather. Meaning the state only represent metrics that have been successfully added to the accumulator.

Docker_log - Separates out the state to be persisted from the frequently updated local state that is updated during the gather. Meaning that again represent metrics that have been successfully added to the accumulator provided in the gather.

Dedup - updates the state during the apply.

Is this a good direction? 

The edge case I found with the current implementation (specifically to the tail plugin) is that if the process never safely shutdowns, then no state is ever recorded. A crash may cause a whole bunch of messages to be lost, since it starts from the end of the file if no state is found.

If this approach is not likely to be wanted, would a better alternative be a option in the tail plugin to read from the beginning if no state is found. Such as 'read_from_beginning_if_no_state'?

**Comment by dayvar14** [comment]
I did notice @srebhan mentioned the following `always returning the state before the actual gather`

> One possible solution would be double buffering. It will not fix the lost-metric example above, but it might guarantee we do have a clear gather-state e.g. by always returning the state before the actual gather...

I might need to look through the code a bit more, but (for non-service plugins) perhaps only persisting the state of previous "gather" during a gather might be better? Maybe that is what was meant.

I'm not currently sure if another gather confirms that the previous gather's metrics were written atm. 

As for the tail plugin, confirming the metrics were delivered before state is written might be a bit more complex.

**Comment by powersj** [comment]
> I might need to look through the code a bit more, but (for non-service plugins) perhaps only persisting the state of previous "gather" during a gather might be better? Maybe that is what was meant.
> I'm not currently sure if another gather confirms that the previous gather's metrics were written atm.

While rare, some cases may involve where a use gathers every 5 seconds, and flushes every 10 seconds. So multiple gathers could occur before a flush happens. But keep in mind this only applies to non-service plugins.

> As for the tail plugin, confirming the metrics were delivered before state is written might be a bit more complex.

100% - this is why he and I are hesitating :) the service plugins (e.g. tail, *-consumer) use tracking metrics are read not at gathers, but instead as they come in and as we have room for more.

**Comment by dayvar14** [comment]
This is fine in our use case, but i can see why it wont be desirable in other situations. 

> A crash may cause a whole bunch of messages to be lost, since it starts from the end of the file if no state is found.

Perhaps a more acceptable solution might be to configure input.tail to read from start of file if no state within statefile is found? Perhaps a `start_from_beginning_if_no_state`? This would lead to duplicates, but those would be more desirable in our situation then dropped logs.

**Comment by srebhan** [comment]
@dayvar14 sorry for the late response! 

> Darn. This is fine in our use case, but i can see why it wont be desirable in most situations.

Yeah and I'm sorry for not being able to offer a good solution! I wish we could provide an easy way...

> The edge case I found with the current main implementation (specifically to the tail plugin) is that if the process never safely shutdowns, then no state is ever recorded. 

This should really be fixed as not shutting down safely is causing other troubles as well. Are you able to reproduce the issue with the latest version? If so, please open a bug-report and let me know!

> Perhaps a more acceptable solution might be to configure input.tail to read from start of file if no state within statefile is found? Perhaps a start_from_beginning_if_no_state option or allow start_from_beginning to work with state?

This is definitively feasible and I think can be done. Instead of introducing a new boolean I would suggest to have
```toml
  ## Read strategy
  ## Available settings are:
  ##   continue      -- continue from the last saved state or, if no state provided, from the end of the file
  ##   from_tail     -- always start from the end of the file no matter if a state is provided or not
  ##   from_head     -- always start from the beginning of the no matter if a state is provided or not
  ##   fallback_head -- continue from the last saved state or, if no state provided, from the start of the file
  # read_strategy = "continue"
```
with `continue` being the default. You might want to improve the naming as this is not my strong part. ;-)

What do you think?

**Comment by dayvar14** [comment]
> This should really be fixed as not shutting down safely is causing other troubles as well. Are you able to reproduce the issue with the latest version? If so, please open a bug-report and let me know!

I believe this is talked about in https://github.com/influxdata/telegraf/issues/14566. State is never recorded as its only recorded when telegraf safely shutdowns.

> This is definitively feasible and I think can be done. Instead of introducing a new boolean I would suggest to have

In that case I will open a new issue with this more specific issue. I'll leave this issue open as i do believe an incrementing state is desirable. Other log agents like fluentd seem to have this feature and in our use case it is important minimize the amount of our critical logs lost.

> with continue being the default. You might want to improve the naming as this is not my strong part. ;-)

Naming is always hard!

**Comment by srebhan** [comment]
@dayvar14 can we please open an issue for this with the _exact_ requirements and close the PR? We can discuss ways forward in the issue... Keeping the PR open will only clutter the list and makes it harder for us to maintain an overview of what is open...

**Pull request closed without merging**